File: apps/api/src/tip_api/contracts/analytics/v1/review_deployment.py
"""Exact, one-release authorization contract for a stale Production review."""

from __future__ import annotations

from datetime import date
from typing import Annotated, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, model_validator


REVIEW_DEPLOYMENT_CONTRACT_VERSION = "production-review-deployment/1.0"
REVIEW_APPROVED_AS_OF_SESSION = date(2026, 8, 24)
REVIEW_EXPECTED_LATEST_SESSION = date(2026, 8, 25)
REVIEW_EXPECTED_LAG_SESSIONS = 1
REVIEW_ACKNOWLEDGEMENT = "I_ACKNOWLEDGE_2026_08_24_STALE_REVIEW_LAG_1"
REVIEW_DEPLOYMENT_CONTRACT_VERSION_V1_1 = "production-review-deployment/1.1"
REVIEW_APPROVED_AS_OF_SESSION_V1_1 = date(2026, 8, 26)
REVIEW_EXPECTED_LATEST_SESSION_V1_1 = date(2026, 8, 27)
REVIEW_EXPECTED_LAG_SESSIONS_V1_1 = 1
REVIEW_ACKNOWLEDGEMENT_V1_1 = "I_ACKNOWLEDGE_2026_08_26_STALE_REVIEW_LAG_1"


class ReviewDeploymentAuthorizationV1(BaseModel):
    """A deliberately non-general authorization for the approved 2026-08-24 review."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    contract_version: Literal["production-review-deployment/1.0"] = (
        REVIEW_DEPLOYMENT_CONTRACT_VERSION
    )
    review_mode: Literal[True] = True
    normal_freshness: Literal[False] = False
    data_status: Literal["stale_review"] = "stale_review"
    approved_as_of_session: date
    expected_latest_session: date
    expected_lag_sessions: Literal[1]
    explicit_user_acknowledgement: str

    @model_validator(mode="after")
    def exact_authorization(self) -> "ReviewDeploymentAuthorizationV1":
        if (
            self.approved_as_of_session != REVIEW_APPROVED_AS_OF_SESSION
            or self.expected_latest_session != REVIEW_EXPECTED_LATEST_SESSION
            or self.expected_lag_sessions != REVIEW_EXPECTED_LAG_SESSIONS
            or self.explicit_user_acknowledgement != REVIEW_ACKNOWLEDGEMENT
        ):
            raise ValueError("review deployment authorization is not the approved exact binding")
        return self


class ReviewDeploymentAuthorizationV1_1(BaseModel):
    """A deliberately non-general authorization for the approved 2026-08-26 review."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    contract_version: Literal["production-review-deployment/1.1"] = (
        REVIEW_DEPLOYMENT_CONTRACT_VERSION_V1_1
    )
    review_mode: Literal[True] = True
    normal_freshness: Literal[False] = False
    data_status: Literal["stale_review"] = "stale_review"
    approved_as_of_session: date
    expected_latest_session: date
    expected_lag_sessions: Literal[1]
    explicit_user_acknowledgement: str

    @model_validator(mode="after")
    def exact_authorization(self) -> "ReviewDeploymentAuthorizationV1_1":
        if (
            self.approved_as_of_session != REVIEW_APPROVED_AS_OF_SESSION_V1_1
            or self.expected_latest_session != REVIEW_EXPECTED_LATEST_SESSION_V1_1
            or self.expected_lag_sessions != REVIEW_EXPECTED_LAG_SESSIONS_V1_1
            or self.explicit_user_acknowledgement != REVIEW_ACKNOWLEDGEMENT_V1_1
        ):
            raise ValueError("review deployment authorization is not the approved exact binding")
        return self


ReviewDeploymentAuthorization: TypeAlias = Annotated[
    ReviewDeploymentAuthorizationV1 | ReviewDeploymentAuthorizationV1_1,
    Field(discriminator="contract_version"),
]


def approved_review_authorization(
    *,
    approved_as_of_session: date,
    expected_latest_session: date,
    expected_lag_sessions: int,
    explicit_user_acknowledgement: str,
) -> ReviewDeploymentAuthorization:
    values = {
        "approved_as_of_session": approved_as_of_session,
        "expected_latest_session": expected_latest_session,
        "expected_lag_sessions": expected_lag_sessions,
        "explicit_user_acknowledgement": explicit_user_acknowledgement,
    }
    if explicit_user_acknowledgement == REVIEW_ACKNOWLEDGEMENT:
        return ReviewDeploymentAuthorizationV1(**values)
    if explicit_user_acknowledgement == REVIEW_ACKNOWLEDGEMENT_V1_1:
        return ReviewDeploymentAuthorizationV1_1(**values)
    raise ValueError("review deployment authorization is not the approved exact binding")


def review_acknowledgement_for_contract(contract_version: str) -> str:
    """Return the exact acknowledgement bound to one supported review contract."""

    if contract_version == REVIEW_DEPLOYMENT_CONTRACT_VERSION:
        return REVIEW_ACKNOWLEDGEMENT
    if contract_version == REVIEW_DEPLOYMENT_CONTRACT_VERSION_V1_1:
        return REVIEW_ACKNOWLEDGEMENT_V1_1
    raise ValueError("unsupported review deployment contract")

File: apps/web/src/api/reviewDeployment.ts
export type ReviewDeploymentContractVersion =
  | 'production-review-deployment/1.0'
  | 'production-review-deployment/1.1';

export interface ReviewDeployment {
  contract_version: ReviewDeploymentContractVersion;
  review_mode: true;
  normal_freshness: false;
  data_status: 'stale_review';
  approved_as_of_session: string;
  expected_latest_session: string;
  expected_lag_sessions: 1;
  explicit_user_acknowledgement: string;
}

const BINDINGS: Record<ReviewDeploymentContractVersion, {
  approved: string; expected: string; acknowledgement: string;
}> = {
  'production-review-deployment/1.0': {
    approved: '2026-08-24', expected: '2026-08-25',
    acknowledgement: 'I_ACKNOWLEDGE_2026_08_24_STALE_REVIEW_LAG_1',
  },
  'production-review-deployment/1.1': {
    approved: '2026-08-26', expected: '2026-08-27',
    acknowledgement: 'I_ACKNOWLEDGE_2026_08_26_STALE_REVIEW_LAG_1',
  },
};

export function reviewMetadataMatches(
  contract: unknown, approved: unknown, expected: unknown, lag: unknown,
): boolean {
  if (typeof contract !== 'string' || !(contract in BINDINGS)) return false;
  const binding = BINDINGS[contract as ReviewDeploymentContractVersion];
  return approved === binding.approved && expected === binding.expected && lag === 1;
}

export function parseReviewDeployment(value: unknown, asOfSession: unknown): ReviewDeployment {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error('Invalid review deployment contract');
  }
  const review = value as Record<string, unknown>;
  if (!reviewMetadataMatches(
    review.contract_version, review.approved_as_of_session,
    review.expected_latest_session, review.expected_lag_sessions,
  )) throw new Error('Invalid review deployment contract');
  const binding = BINDINGS[review.contract_version as ReviewDeploymentContractVersion];
  if (review.review_mode !== true || review.normal_freshness !== false
    || review.data_status !== 'stale_review'
    || review.approved_as_of_session !== asOfSession
    || review.explicit_user_acknowledgement !== binding.acknowledgement) {
    throw new Error('Invalid review deployment contract');
  }
  return review as unknown as ReviewDeployment;
}
